File: sdks/python/src/bcs/uleb128.py
"""ULEB128 encoding and decoding for BCS."""

from __future__ import annotations

from .errors import NonCanonicalUleb128, Uleb128Overflow, UnexpectedEof

# Maximum value that can be encoded (u32 max)
MAX_U32 = 0xFFFFFFFF


def encode(value: int) -> bytes:
    """Encode an unsigned integer as ULEB128.

    Args:
        value: Non-negative integer to encode (must fit in u32)

    Returns:
        ULEB128 encoded bytes

    Raises:
        ValueError: If value is negative or exceeds u32 max
    """
    if value < 0:
        raise ValueError(f"ULEB128 cannot encode negative values: {value}")
    if value > MAX_U32:
        raise ValueError(f"ULEB128 value exceeds u32 max: {value}")

    result = bytearray()
    while value >= 0x80:
        # Write 7 bits with continuation bit set
        result.append((value & 0x7F) | 0x80)
        value >>= 7
    # Write final byte without continuation bit
    result.append(value)
    return bytes(result)


def decode(data: bytes | bytearray | memoryview, offset: int = 0) -> tuple[int, int]:
    """Decode a ULEB128 value from bytes.

    Args:
        data: Bytes to decode from
        offset: Starting position in data

    Returns:
        Tuple of (decoded value, new offset after reading)

    Raises:
        UnexpectedEof: If data ends before ULEB128 is complete
        NonCanonicalUleb128: If encoding is not minimal (has trailing zeros)
        Uleb128Overflow: If value exceeds u32 max or encoding is too long
    """
    value = 0
    shift = 0

    for i in range(5):  # Maximum 5 bytes for u32
        if offset + i >= len(data):
            raise UnexpectedEof()

        byte = data[offset + i]
        digit = byte & 0x7F
        value |= digit << shift

        # Check if this is the last byte (high bit is 0)
        if digit == byte:
            # Reject non-canonical encodings (trailing zero bytes)
            if shift > 0 and digit == 0:
                raise NonCanonicalUleb128()

            # Check for overflow
            if value > MAX_U32:
                raise Uleb128Overflow()

            return value, offset + i + 1

        shift += 7

    # If we get here, we read 5 bytes and all had continuation bits
    raise Uleb128Overflow()


def encoded_size(value: int) -> int:
    """Calculate the number of bytes needed to encode a value.

    Args:
        value: Non-negative integer

    Returns:
        Number of bytes required for ULEB128 encoding
    """
    if value == 0:
        return 1
    size = 0
    while value > 0:
        size += 1
        value >>= 7
    return size

File: sdks/elixir/lib/bcs/uleb128.ex
defmodule Bcs.Uleb128 do
  @moduledoc """
  ULEB128 encoding and decoding for BCS.

  ULEB128 (Unsigned Little-Endian Base 128) is a variable-length encoding
  for unsigned integers. Each byte contributes 7 bits of data, with the
  high bit indicating whether more bytes follow.
  """

  alias Bcs.Error

  # Maximum value that can be encoded (u32 max)
  @max_u32 0xFFFFFFFF

  @doc """
  Encode an unsigned integer as ULEB128.

  ## Examples

      iex> Bcs.Uleb128.encode(0)
      <<0x00>>

      iex> Bcs.Uleb128.encode(127)
      <<0x7F>>

      iex> Bcs.Uleb128.encode(128)
      <<0x80, 0x01>>

      iex> Bcs.Uleb128.encode(16384)
      <<0x80, 0x80, 0x01>>

  """
  @spec encode(non_neg_integer()) :: binary()
  def encode(value) when is_integer(value) and value >= 0 and value <= @max_u32 do
    do_encode(value, <<>>)
  end

  def encode(value) when is_integer(value) and value < 0 do
    raise ArgumentError, "ULEB128 cannot encode negative values: #{value}"
  end

  def encode(value) when is_integer(value) do
    raise ArgumentError, "ULEB128 value exceeds u32 max: #{value}"
  end

  defp do_encode(value, acc) when value < 0x80 do
    acc <> <<value>>
  end

  defp do_encode(value, acc) do
    byte = Bitwise.band(value, 0x7F) |> Bitwise.bor(0x80)
    do_encode(Bitwise.bsr(value, 7), acc <> <<byte>>)
  end

  @doc """
  Decode a ULEB128 value from binary data.

  Returns `{:ok, {value, rest}}` on success, or `{:error, reason}` on failure.

  ## Examples

      iex> Bcs.Uleb128.decode(<<0x00>>)
      {:ok, {0, <<>>}}

      iex> Bcs.Uleb128.decode(<<0x7F>>)
      {:ok, {127, <<>>}}

      iex> Bcs.Uleb128.decode(<<0x80, 0x01>>)
      {:ok, {128, <<>>}}

      iex> Bcs.Uleb128.decode(<<0x80, 0x01, 0xFF>>)
      {:ok, {128, <<0xFF>>}}

  """
  @spec decode(binary()) :: {:ok, {non_neg_integer(), binary()}} | {:error, Error.t()}
  def decode(data) when is_binary(data) do
    do_decode(data, 0, 0, 0)
  end

  # Successfully decoded - check for non-canonical encoding
  defp do_decode(<<byte, rest::binary>>, value, shift, _count) when byte < 0x80 do
    final_value = Bitwise.bor(value, Bitwise.bsl(byte, shift))

    cond do
      # Non-canonical: trailing zero byte (except for value 0 at position 0)
      shift > 0 and byte == 0 ->
        {:error, Error.non_canonical_uleb128()}

      # Overflow check
      final_value > @max_u32 ->
        {:error, Error.uleb128_overflow()}

      true ->
        {:ok, {final_value, rest}}
    end
  end

  # Continue decoding
  defp do_decode(<<byte, rest::binary>>, value, shift, count) when count < 4 do
    digit = Bitwise.band(byte, 0x7F)
    new_value = Bitwise.bor(value, Bitwise.bsl(digit, shift))
    do_decode(rest, new_value, shift + 7, count + 1)
  end

  # 5th byte - must be final and small enough
  defp do_decode(<<byte, rest::binary>>, value, shift, 4) when byte < 0x10 do
    final_value = Bitwise.bor(value, Bitwise.bsl(byte, shift))

    if final_value > @max_u32 do
      {:error, Error.uleb128_overflow()}
    else
      {:ok, {final_value, rest}}
    end
  end

  # 5th byte with continuation bit or too large
  defp do_decode(<<_byte, _rest::binary>>, _value, _shift, 4) do
    {:error, Error.uleb128_overflow()}
  end

  # Unexpected EOF
  defp do_decode(<<>>, _value, _shift, _count) do
    {:error, Error.unexpected_eof()}
  end

  @doc """
  Decode a ULEB128 value, raising on error.

  ## Examples

      iex> Bcs.Uleb128.decode!(<<0x80, 0x01>>)
      {128, <<>>}

  """
  @spec decode!(binary()) :: {non_neg_integer(), binary()}
  def decode!(data) do
    case decode(data) do
      {:ok, result} -> result
      {:error, error} -> raise error
    end
  end

  @doc """
  Calculate the number of bytes needed to encode a value.

  ## Examples

      iex> Bcs.Uleb128.encoded_size(0)
      1

      iex> Bcs.Uleb128.encoded_size(127)
      1

      iex> Bcs.Uleb128.encoded_size(128)
      2

      iex> Bcs.Uleb128.encoded_size(16384)
      3

  """
  @spec encoded_size(non_neg_integer()) :: pos_integer()
  def encoded_size(0), do: 1

  def encoded_size(value) when is_integer(value) and value > 0 do
    do_encoded_size(value, 0)
  end

  defp do_encoded_size(0, count), do: count
  defp do_encoded_size(value, count), do: do_encoded_size(Bitwise.bsr(value, 7), count + 1)
end
